### PR TITLE
feat (robot model): Damped least squared pseudoinverse

### PR DESCRIPTION
**Pull request by JasperTan97**
## Description

When robots are moving, the joints can start to move toward a singular configuration. As the rank of the jacobian start to become more deficient, numerical methods can select a very high joint velocity to solve the inverse kinematics problem. Here, a damped least squared pseudo inverse method is applied to the controller where the inverse kinematics equation is solved with a ridge regression term with coefficient dls_lambda. This solution trades away optimality with practical joint speeds.

Furthermore, a unit test and the corresponding generating matlab code was updated to test the correctness of the method. Currently, the code passed all three randomly generated tests. Furthermore, we see below that the damping does in fact reduce joint speeds in a near singular configuration.

It must be noted that this solution does not solve the problem where the robot is in a singular configuration. In this state, the IK solver will still pick (near) infinite joint speeds. Secondly, joint speeds can still become very high. This solution tries to keep the 2-norm of the joint speeds down. This doesn't work if the requirement of the IK solution outweighs the damping term. A hard limit is required to solve this problem.

Finally, this PR does not test that the controller is working as intended, only that the joint speeds are decreased. For that, we should discuss what is a suitable test for correctness of solution. At least, I would like to hook the controller up to the simulator, and in rviz, check the robot still moves to the correct location (since we already know that joint speeds decrease). For this, we will have to edit the backend files like adding a new parameter (dls_lambda) into the velocity_control.yaml. (anything else?)

## Supporting information
Screenshot from 2023-09-25 12-44-29

Figure 1: Joint speeds goes down!

## Review guidelines

Estimated Time of Review: 20 minutes

## Checklist before merging:

- [x] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes
- [x] Tested on mujoco simulator with rviz interactive marker

### CODE REVIEW

**Comment by JasperTan97** [comment]
I have read and understood the Contributor License Agreement and hereby sign it

File: source/robot_model/src/Model.cpp [hunk 1]
```diff
@@ -411,6 +412,17 @@ Model::inverse_velocity(const std::vector<state_representation::CartesianTwist>&
...
 
+  // add damped least square term
+  if (jacobian.rows() > jacobian.cols()){
+    jacobian = jacobian.transpose().completeOrthogonalDecomposition().pseudoInverse() * 
+                (jacobian.transpose() * jacobian + 
+                dls_lambda * dls_lambda * Eigen::MatrixXd::Identity(jacobian.cols(), jacobian.cols()));
+  } else {
+    jacobian = (jacobian * jacobian.transpose() + 
+                dls_lambda * dls_lambda * Eigen::MatrixXd::Identity(jacobian.rows(), jacobian.rows())) *
+                jacobian.transpose().completeOrthogonalDecomposition().pseudoInverse();
```
Comment (eeberhard) [inline source/robot_model/src/Model.cpp:423]: This looks incorrect to me. The first block $JJ^T + \lambda^2 I$ will be a square matrix with size `jacobian.rows()`. But then you multiply it by the Jacobian transpose, so for any non-square Jacobian you'll have an incompatible matrix product.

i.e. if the Jacobian is of size 6x7 (very common case with 7-joint robots), then this operation would be 6x6 * 7x6

This is based on your choice of `if (jacobian.rows() > jacobian.cols())` above, where this branch includes any Jacobians where `jacobian.rows() <= jacobian.cols()`

**Review by eeberhard — COMMENTED** [review]
The damped version of the pseudo-inverse (let's call it $\hat{J}^+$) is usually formulated as:

$$
\hat{J}^+ = J^T(J J^T + \lambda^2 I)^{-1}
$$

As such, it can directly be used to solve for joint velocities as $\dot{\theta} = \hat{J}^+ \dot{x}$.

Now as you have observed, the existing IK function uses QR decomposition to solve for joint velocities $\dot{\theta}$ in the linear system $J\dot{\theta} = \dot{x}$, which is generally more efficient than calculating the inverse / pseudo-inverse and then multiplying.

It looks like you have tried to combine these two approaches into one by inverting the equation for `\hat{J}^+` to get back a "non-inverted" but damped jacobian.

For one, I think there might be some issues with your inversion technique. Generally rules like $(AB)^{-1} = B^{-1}A^{-1}$ only hold for square matrices. I'm not enough of a linear algebra guru to know when exceptions can be made.

But overall, I think you end up overcomplicating the code for no benefit. Your inverted formulation has basically the same number of matrix products and inversions as the standard formulation, but then you still pass it to QR decomposition for solving.

To put it simply, I would add check for `if (dls_lambda > 0.0)`, which if true, returns the result of the textbook $\dot{\theta} = \hat{J}^+ \dot{x}$. Else, if there is no damping, we use the existing decomposition solver for efficiency.

The check to use the "right pseudo-inverse" for cases where the joint-space degrees of freedom are equal to or larger than the Cartesian degrees of freedom, and to use the special left-hand case for under-articulated systems is good to keep in there but your inversions of the two cases just looked suspect to me.

File: source/robot_model/src/Model.cpp (hunk 1 above)
Comment (JasperTan97) [reply source/robot_model/src/Model.cpp:423]: > This looks incorrect to me. The first block will be a square matrix with size jacobian.rows(). But then you multiply it by the Jacobian transpose, so for any non-square Jacobian you'll have an incompatible matrix product.

Lastly, I looked at this realised the issue. I am not multiplying with the jacobian transpose but the jacobian transpose pseudoinverse. Here I have some scratchings of why I think the maths is fine.

image

Comment (eeberhard) [reply source/robot_model/src/Model.cpp]: Ah ok, very good point, I forgot that the pseudo inverse also acts as a transposition. Thanks for following up!

File: source/robot_model/src/Model.cpp [hunk 2]
```diff
@@ -411,6 +412,27 @@ Model::inverse_velocity(const std::vector<state_representation::CartesianTwist>&
...
+  } else {
+    Eigen::MatrixXd j_prime = jacobian * jacobian.transpose() + 
+                dls_lambda * dls_lambda * Eigen::MatrixXd::Identity(jacobian.rows(), jacobian.rows());
+    return state_representation::JointVelocities(joint_positions.get_name(),
+                                               joint_positions.get_names(),
+                                               jacobian.transpose() * j_prime.colPivHouseholderQr().solve(dX));
+  }
+
   // solve a linear system
   return state_representation::JointVelocities(joint_positions.get_name(),
```
Comment (eeberhard) [inline source/robot_model/src/Model.cpp:437]: this code is now unreachable, you should remove it

File: source/robot_model/include/robot_model/Model.hpp [hunk 3]
```diff
@@ -419,11 +421,13 @@ class Model {
    * @param frame name of the frame at which to compute the twist
    * @param parameters parameters of the inverse velocity kinematics algorithm (default is default values of the
    * QPInverseVelocityParameters structure)
+   * @param damped least square term
```
Comment (eeberhard) [inline source/robot_model/include/robot_model/Model.hpp:424]:
```suggestion
   * @param dls_lambda damped least square term
```

**Review by eeberhard — COMMENTED** [review]
Thanks for the updates to revise the matrix math. As a whole it looks very efficient now! Also nice initiative to add the test for the 6dof robot too, really appreciate that.

Minor touchups left but other than that it's very good work.

**Review by eeberhard — APPROVED** [review]
Thanks for your patience with the review passes. This looks great and is a nice addition to the base IK solver. Congratulations on your first contribution to CL!
